### PR TITLE
Entity Effect Events

### PR DESCRIPTION
**Pull request by JackNoordhuis**
Add EntityEffectEvent class that the EntityEffectAddEvent and
EntityEffectRemoveEvent classes extend. Add event calls to Effect class.

This is purely an API addition that allows developers to modify the effect behaviour through the use of plugins.

I have tested this and it works as expected. Some more testing may need to be done to make sure the event's work as expect in all cases.

### CODE REVIEW

**Comment by JackNoordhuis** [comment]
This might be useful to anyone wanting to test out these changes: https://github.com/JackNoordhuis/PocketMine-Plugins/tree/master/EffectEventTesting

File: src/pocketmine/entity/Effect.php [hunk 1]
```diff
@@ -251,14 +253,19 @@ public function setColor($r, $g, $b){
 	}
 
 	public function add(Entity $entity, $modify = false, Effect $oldEffect = null){
+		$entity->getLevel()->getServer()->getPluginManager()->callEvent($ev = new EntityEffectAddEvent
+		($entity, $this, $modify, $oldEffect));
```
Comment (dktapps) [inline src/pocketmine/entity/Effect.php]: please fix your formatting here.

Comment (JackNoordhuis) [inline src/pocketmine/entity/Effect.php]: Accidently changed some settings in PHPStorm when I was messing around the other day, fixed.

**Review by dktapps — APPROVED** [review]
Looks good to me, but will let someone else take a look also.

**Comment by JackNoordhuis** [comment]
Squashed commits just because a derpy formatting fix doesn't deserve its own commit and for convenience.

File: src/pocketmine/event/entity/EntityEffectAddEvent.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	/** @var Effect */
+	private $oldEffect;
+
+	public function __construct(Entity $entity, Effect $effect, $modify, $oldEffect){
+		parent::__construct($entity, $effect);
+		$this->modify = $modify;
+		$this->oldEffect = $oldEffect;
+	}
+
+	public function willModify(): bool{
```
Comment (SOF3) [inline src/pocketmine/event/entity/EntityEffectAddEvent.php]: I thought our code style is `willModify() : bool{`

**Review by SOF3 — CHANGES_REQUESTED** [review]

File: src/pocketmine/event/entity/EntityEffectAddEvent.php (hunk 2 above)
Comment (dktapps) [inline src/pocketmine/event/entity/EntityEffectAddEvent.php]: OCD ftw...

File: src/pocketmine/event/entity/EntityEffectAddEvent.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		return $this->modify;
+	}
+
+	public function setWillModify(bool $modify){
+		$this->modify = $modify;
+	}
+
+	public function hasOldEffect(): bool{
```
Comment (dktapps) [inline src/pocketmine/event/entity/EntityEffectAddEvent.php]: @SOF3 missed one :P

File: src/pocketmine/event/entity/EntityEffectEvent.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+	/** @var Effect */
+	private $effect;
+
+	public function __construct(Entity $entity, Effect $effect){
+		$this->entity = $entity;
+		$this->effect = $effect;
+	}
+
+	public function getEffect(): Effect{
```
Comment (dktapps) [inline src/pocketmine/event/entity/EntityEffectEvent.php]: and here ;)

Comment (SOF3) [inline src/pocketmine/event/entity/EntityEffectEvent.php]: I don't need to triple comment these, do I ;)

Comment (dktapps) [inline src/pocketmine/event/entity/EntityEffectEvent.php]: maybe? he won't miss any of them now :O

File: src/pocketmine/event/entity/EntityEffectAddEvent.php (hunk 2 above)
Comment (JackNoordhuis) [inline src/pocketmine/event/entity/EntityEffectAddEvent.php]: Had a feeling it was supposed to be like that, will change.

**Comment by JackNoordhuis** [comment]
Fixed the return type hint formatting.

**Review by SOF3 — APPROVED** [review]

**Pull request merged**